### PR TITLE
add 3dmol.js post

### PR DESCRIPTION
**Pull request by duerrsimon**
I wrote a short post on how to visualize proteins using gradio as discussed with @osanseviero 

Let me know what you think :)

### CODE REVIEW

**Comment by osanseviero** [comment]
As discussed in a DM, we'll review it here first and once ready we'll have a Gradio Guide with same content. I will review this Friday of next week but maybe @abidlabs is interested in reviewing as well! :partying_face: 

Thank you for the contribution!!

File: visualize_proteins_3dmol.js.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
+---
+title: "Visualize proteins on huggingface spaces"
+thumbnail: /blog/assets/94_visualize_proteins_using_3dmoljs/thumbnail.png
+---
+
+<h1>Visualize proteins on Huggingface spaces</h1>
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:6]:
```suggestion
<h1>Visualize proteins on Hugging Face Spaces</h1>

```

File: visualize_proteins_3dmol.js.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
+---
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:1]: You'll also need to add the blog entry to https://github.com/huggingface/blog/blob/main/_blog.yml for it to be shown

File: visualize_proteins_3dmol.js.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
+---
+title: "Visualize proteins on huggingface spaces"
+thumbnail: /blog/assets/94_visualize_proteins_using_3dmoljs/thumbnail.png
+---
+
+<h1>Visualize proteins on Huggingface spaces</h1>
+
+<div class="blog-metadata">
+    <small>Published 10.08, 2022.</small>
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:9]:
```suggestion
    <small>Published August 24, 2022.</small>
```

File: visualize_proteins_3dmol.js.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
+---
+title: "Visualize proteins on huggingface spaces"
+thumbnail: /blog/assets/94_visualize_proteins_using_3dmoljs/thumbnail.png
+---
+
+<h1>Visualize proteins on Huggingface spaces</h1>
+
+<div class="blog-metadata">
+    <small>Published 10.08, 2022.</small>
+    <a target="_blank" class="btn no-underline text-sm mb-5 font-sans" href="https://github.com/huggingface/blog/blob/main/pytorch-fsdp.md">
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:10]: Please update the path to the markdown

File: visualize_proteins_3dmol.js.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    <a href="/duerrsimon">
+        <img class="avatar avatar-user" src="hhttps://avatars.githubusercontent.com/u/16979577?s=200&v=4" title="Gravatar">
+        <div class="bfc">
+            <code>duerrsimon</code>
+            <span class="fullname">Simon Dürr</span>
+        </div>
+    </a>
+</div>
+
+In this post we will look at how we can visualize proteins on Huggingface spaces.
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:25]:
```suggestion
In this post we will look at how we can visualize proteins on Hugging Face Spaces.

```

File: visualize_proteins_3dmol.js.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        Update on GitHub
+    </a>
+</div>
+
+<div class="author-card">
+    <a href="/duerrsimon">
+        <img class="avatar avatar-user" src="hhttps://avatars.githubusercontent.com/u/16979577?s=200&v=4" title="Gravatar">
+        <div class="bfc">
+            <code>duerrsimon</code>
+            <span class="fullname">Simon Dürr</span>
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:20]: Could you add a guest tag as in https://github.com/merveenoyan/blog/blob/9fac04883b50f948e75065e51d1396bdc23bbfa4/deploy-tfserving-kubernetes.md?plain=1#L31?

File: visualize_proteins_3dmol.js.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+A famous ML model for proteins is AlphaFold2 which predicts the structure of a protein sequence using a multiple sequence alignment of similar proteins and a structure module. 
+
+Since AlphaFold2 made its debut many more such models have come out such as OmegaFold, OpenFold etc. (see this [list](https://github.com/yangkky/Machine-learning-for-proteins) or this [list](https://github.com/sacdallago/folding_tools) for more). 
+
+
+## Seeing is believing
+
+The structure of a protein is an integral part to our understanding of what a protein does. Nowadays, there are a few tools available to visualize proteins directly in the browser such as [mol*](molstar.org) or [3dmol.js](https://3dmol.csb.pitt.edu/). In this post, you will learn how to integrate structure visualization into your Huggingface space using 3Dmol.js and the HTML block.
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:38]:
```suggestion
The structure of a protein is an integral part to our understanding of what a protein does. Nowadays, there are a few tools available to visualize proteins directly in the browser such as [mol*](molstar.org) or [3dmol.js](https://3dmol.csb.pitt.edu/). In this post, you will learn how to integrate structure visualization into your Hugging Face Space using 3Dmol.js and the HTML block. 

```

File: visualize_proteins_3dmol.js.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+## Prerequisites
+
+Make sure you have the `gradio` Python package already [installed](/getting_started) and basic knowledge of Javascript/JQuery.
+
+
+## Taking a Look at the Code
+
+Let's take a look at how to create the minimal working demo of our interface before we dive into how to setup 3Dmol.js. 
+
+We will built a simple demo app that can accept either a 4-digit PDB code  or a PDB file. Our app will then retrieve the pdb file from the RCSB Protein Databank and display it or use the uploaded file for display.
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:49]:
```suggestion
We will build a simple demo app that can accept either a 4-digit PDB code or a PDB file. Our app will then retrieve the pdb file from the RCSB Protein Databank and display it or use the uploaded file for display.

```

File: visualize_proteins_3dmol.js.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+<iframe src="https://hf.space/embed/simonduerr/3dmol.js/+
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:51]: Could we embed the space with https://gradio.app/sharing_your_app/#embedding-with-web-components? It has the benefit of being responsive and nicer code :) You can add the script part here

File: visualize_proteins_3dmol.js.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+" frameBorder="0" width="1400" height="690" title="Gradio app" class="p-0 flex-grow space-iframe" allow="accelerometer; ambient-light-sensor; autoplay; battery; camera; document-domain; encrypted-media; fullscreen; geolocation; gyroscope; layout-animations; legacy-image-formats; magnetometer; microphone; midi; oversized-images; payment; picture-in-picture; publickey-credentials-get; sync-xhr; usb; vr ; wake-lock; xr-spatial-tracking" sandbox="allow-forms allow-modals allow-popups allow-popups-to-escape-sandbox allow-same-origin allow-scripts allow-downloads"></iframe>
+
+```python
+import gradio as gr
+
+def update(inp, file):
+    # in this simple example we just retrieve the pdb file using its identifier from the RCSB or display the uploaded  file
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:58]:
```suggestion
    # in this simple example we just retrieve the pdb file using its identifier from the RCSB or display the uploaded file

```

File: visualize_proteins_3dmol.js.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+We use an `iframe` for this. 
+
+Our `molecule` function which returns the `iframe` conceptually looks like this: 
+
+```python
+def molecule(input_pdb):
+    mol = read_mol(input_pdb)
+    # setup HTML document
+    x = ("""<!DOCTYPE html><html> [..] </html>""") # do not use ' in this input
+    return f"""<iframe  [..] srcdoc='{x}'></iframe>
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:106]: Starting this line I see lots of `^M` characters. Is this something we could remove?

File: visualize_proteins_3dmol.js.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+            viewer.zoomTo();
+            viewer.render();
+            viewer.zoom(0.8, 2000);
+            })
+    </script>
+</body>
+```
+We use a template literal (denoted by backticks) to store our pdb file in the html document directly and then output it using 3dmol.js.
+
+And that's it, now you can couple your favorite protein ML model to a fun and easy to use gradio app and directly visualize predicted or redesigned structures. If you are predicting properities of a structure (e.g how likely each amino acid is to bind a ligand) 3Dmol.js also allows to use custom `colorfunc` based on a property of each atom.
```
Comment (osanseviero) [inline visualize_proteins_3dmol.js.md:154]:
```suggestion
And that's it, now you can couple your favorite protein ML model to a fun and easy to use gradio app and directly visualize predicted or redesigned structures. If you are predicting properities of a structure (e.g how likely each amino acid is to bind a ligand), 3Dmol.js also allows to use a custom `colorfunc` based on a property of each atom. 

```

**Review by osanseviero — APPROVED** [review]
Very cool blog post! Thanks a lot for writing it! I did a first pass but this is already looking in a great direction :fire:

File: visualize_proteins_3dmol.js.md (hunk 12 above)
Comment (duerrsimon) [reply visualize_proteins_3dmol.js.md:154]: I copy pasted the code on my windows machine and that must have inserted these line endings. I don't see them in VS Code and they only show up in the github soure. I'll get this fixed

File: visualize_proteins_3dmol.js.md (hunk 2 above)
Comment (duerrsimon) [reply visualize_proteins_3dmol.js.md:1]: I used only the research tag. Not sure if you want add one related to protein ML

File: visualize_proteins_3dmol.js.md (hunk 11 above)
Comment (duerrsimon) [reply visualize_proteins_3dmol.js.md]: I fixed it now using dos2unix

**Comment by duerrsimon** [comment]
All requested changes have been incorporated. I also removed the duplicate id for the post (there was already a folder assets/94_xx so I changed it to the first free number (98).

**Comment by osanseviero** [comment]
Awesome, amazing work :) I'll merge+deploy this tomorrow

**Pull request merged**